mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -233,58 +246,65 @@ async fn block_header_utxo_loop<T: UtxoCommonOps>(
             },
         };
 
-        let to_block_height = match coin.as_ref().rpc_client.get_block_count().compat().await {
-            Ok(h) => h,
-            Err(e) => {
-                error!("Error {} on getting the height of the latest block from rpc!", e);
-                sync_status_loop_handle.notify_on_temp_error(e.to_string());
-                Timer::sleep(10.).await;
-                continue;
-            },
-        };
-
         // Todo: Add code for the case if a chain reorganization happens
-        if from_block_height == to_block_height {
+        if from_block_height == last_block_height {
+            sync_status_loop_handle.notify_sync_finished(last_block_height);
+            last_block_height = match coin.as_ref().rpc_client.get_block_count().compat().await {
+                Ok(h) => h,
+                Err(e) => {
+                    error!("Error {} on getting the height of the latest block from rpc!", e);
+                    sync_status_loop_handle.notify_on_temp_error(e.to_string());
+                    Timer::sleep(10.0).await;
+                    continue;
+                },
+            };
             Timer::sleep(BLOCK_HEADERS_LOOP_INTERVAL).await;
             continue;
         }
 
+        let to_block_height = from_block_height + chunk_size;
         sync_status_loop_handle.notify_blocks_headers_sync_status(from_block_height + 1, to_block_height);
 
-        let (block_registry, block_headers, last_retrieved_height) = match client
+        let (block_registry, block_headers) = match client
             .retrieve_headers(from_block_height + 1, to_block_height)
             .compat()
             .await
         {
             Ok(res) => res,
-            Err(e) => {
-                error!("Error {} on retrieving the latest headers from rpc!", e);
-                sync_status_loop_handle.notify_on_temp_error(e.to_string());
-                Timer::sleep(10.).await;
-                continue;
+            Err(error) => {
+                if error.get_inner().is_network_error() {
+                    log!("Network Error: Will try fetching block headers again after 10 secs");
+                    sync_status_loop_handle.notify_on_temp_error(error.to_string());
+                    Timer::sleep(10.).await;
+                    continue;
+                };
+
+                // If electrum returns response too large error, we will request the headers again in chunks of ELECTRUM_MAX_CHUNK_SIZE - CHUNK_SIZE_REDUCER_VALUE instead.
+                if error.get_inner().is_response_too_large() {
+                    chunk_size -= CHUNK_SIZE_REDUCER_VALUE;
+                    continue;
+                }
+
+                error!("Error {} on retrieving the latest headers from rpc!", error);
+                sync_status_loop_handle.notify_on_permanent_error(error.to_string());
+                break;
```

Comment:
You can add `// Todo: remove this electrum server and use another in this case since the headers from this server can't be retrieved` above `sync_status_loop_handle.notify_on_permanent_error(error.to_string());`